tests/src/core/testqgslayoutexporter.cpp

```diff
@@ -0,0 +1,260 @@
+#include "qgslayoutitemlabel.h"
+#include "qgslayoutitemshape.h"
+#include "qgslayoutitemscalebar.h"
+#include "qgslayoutitemmap.h"
+#include "qgsvectorlayer.h"
+#include "qgslayoutitemlegend.h"
+
+class TestQgsLayoutExporter: public QObject
+{
+    Q_OBJECT
+
+  private slots:
+    void initTestCase();// will be called before the first testfunction is executed.
+    void cleanupTestCase();// will be called after the last testfunction was executed.
+    void init();// will be called before each testfunction is executed.
+    void cleanup();// will be called after every testfunction.
+    void testHandleLayeredExport();
+
+};
+
+void TestQgsLayoutExporter::initTestCase()
+{
+  QgsApplication::init();
+  QgsApplication::initQgis();
+}
+
+void TestQgsLayoutExporter::cleanupTestCase()
+{
+}
+
+void TestQgsLayoutExporter::init()
+{
+
+}
+
+void TestQgsLayoutExporter::cleanup()
+{
+
+}
+
+void TestQgsLayoutExporter::testHandleLayeredExport()
+{
+  QgsProject p;
+  QgsLayout l( &p );
+  QgsLayoutExporter exporter( &l );
+
+  QList< unsigned int > layerIds;
+  QStringList layerNames;
+  QStringList mapLayerIds;
+  auto exportFunc = [&layerIds, &layerNames, &mapLayerIds]( unsigned int layerId, const QgsLayoutItem::ExportLayerDetail & layerDetail )->QgsLayoutExporter::ExportResult
+  {
+    layerIds << layerId;
+    layerNames << layerDetail.name;
+    mapLayerIds << layerDetail.mapLayerId;
+    return QgsLayoutExporter::Success;
+  };
+
+  QList< QGraphicsItem * > items;
+  QgsLayoutExporter::ExportResult res = exporter.handleLayeredExport( items, exportFunc );
+  QCOMPARE( res, QgsLayoutExporter::Success );
+  QVERIFY( layerIds.isEmpty() );
+  QVERIFY( layerNames.isEmpty() );
+  QVERIFY( mapLayerIds.isEmpty() );
+
+  // add two pages to a layout
+  QgsLayoutItemPage *page1 = new QgsLayoutItemPage( &l );
+  items << page1;
+  res = exporter.handleLayeredExport( items, exportFunc );
+  QCOMPARE( res, QgsLayoutExporter::Success );
+  QCOMPARE( layerIds, QList< unsigned int >() << 1 );
+  QCOMPARE( layerNames, QStringList() << QStringLiteral( "Page" ) );
+  QCOMPARE( mapLayerIds, QStringList() << QString() );
+  layerIds.clear();
+  layerNames.clear();
+  mapLayerIds.clear();
+
+  QgsLayoutItemPage *page2 = new QgsLayoutItemPage( &l );
+  items << page2;
+  res = exporter.handleLayeredExport( items, exportFunc );
+  QCOMPARE( res, QgsLayoutExporter::Success );
+  QCOMPARE( layerIds, QList< unsigned int >() << 1 );
+  QCOMPARE( layerNames, QStringList() << QStringLiteral( "Pages" ) );
+  QCOMPARE( mapLayerIds, QStringList() << QString() );
+  layerIds.clear();
+  layerNames.clear();
+  mapLayerIds.clear();
+
+  QgsLayoutItemLabel *label = new QgsLayoutItemLabel( &l );
+  items << label;
+  res = exporter.handleLayeredExport( items, exportFunc );
+  QCOMPARE( res, QgsLayoutExporter::Success );
+  QCOMPARE( layerIds, QList< unsigned int >() << 1 << 2 );
+  QCOMPARE( layerNames, QStringList() << QStringLiteral( "Pages" ) << QStringLiteral( "<Label>" ) );
```

Comment:
Is there a reason you enclose Label within lesser than / greater than signs, and not do so for Page(s)? I'd remove the symbols here (you also don't do it when naming layers containing multiple item types, e.g. "Label, Shape").

Reply:
It's coming straight from the layout item "displayName" (i.e. the value shown in the layout items tree widget). By default this is of the format <Scalebar>, <Legend>, etc, UNLESS the item has an ID set, in which case it's the ID. Also, some item types override this, e.g. labels, which return just the first portion of the label's text.

So while in principal I agree that the mix looks messy, I can't think of a clean way to avoid this.

Reply:
Could you clean it up in the nameForLayerWithItems() function, whereas you could strip the first and last chars of a single item layer's displayName of those are < and > ?